zcncore/transaction_query.go

```diff
@@ -0,0 +1,401 @@
+			return thrown.Throw(ErrSharderOffline, resp.Status)
+		}
+
+		return nil
+	})
+	errs := r.Wait()
+
+	if len(errs) > 0 {
+		tq.offline[host] = true
+
+		if len(tq.offline) >= tq.max {
+			return ErrNoOnlineSharders
+		}
+	}
+
+	return nil
+}
+
+// randOne random one health sharder
+func (tq *TransactionQuery) randOne(ctx context.Context) (string, error) {
+
+	randGen := rand.New(rand.NewSource(time.Now().UnixNano()))
+	for {
+
+		// reset selected if all sharders were selected
+		if len(tq.selected) >= tq.max {
+			tq.selected = make(map[string]interface{})
+		}
+
+		i := randGen.Intn(len(tq.sharders))
+		host := tq.sharders[i]
+
+		_, ok := tq.selected[host]
+
+		// it was selected, try next
+		if ok {
+			continue
+		}
+
+		tq.selected[host] = true
+
+		err := tq.checkHealth(ctx, host)
+
+		if err != nil {
+			if errors.Is(err, ErrNoOnlineSharders) {
+				return "", err
+			}
+
+			// it is offline, try next one
+			continue
+		}
+
+		return host, nil
+	}
+}
+
+// FromAny query transaction form any sharder that is not selected in previous queires. use any used sharder if there is not any unused sharder
+func (tq *TransactionQuery) FromAny(ctx context.Context, query string) (QueryResult, error) {
+
+	res := QueryResult{
+		StatusCode: http.StatusBadRequest,
+	}
+
+	err := tq.validate(1)
+
+	if err != nil {
+		return res, err
+	}
+
+	host, err := tq.randOne(ctx)
+
+	if err != nil {
+		return res, err
+	}
+
+	r := resty.New()
+	requestUrl := tq.buildUrl(host, query)
+
+	Logger.Debug("GET", requestUrl)
+
+	r.DoGet(ctx, requestUrl).
+		Then(func(req *http.Request, resp *http.Response, respBody []byte, cf context.CancelFunc, err error) error {
+			Logger.Debug(requestUrl + " " + resp.Status)
+
+			res.Error = err
+			if err != nil {
+				return err
+			}
+
+			res.Content = respBody
+			res.StatusCode = resp.StatusCode
+
+			Logger.Debug(string(respBody))
+
+			return nil
+		})
+
+	errs := r.Wait()
+
+	if len(errs) > 0 {
+		return res, errs[0]
+	}
+
+	return res, nil
+
+}
+
+// FromConsensus query transaction from all sharders whatever it is selected or offline in previous queires, and return consensus result
+func (tq *TransactionQuery) FromConsensus(ctx context.Context, query string, handle QueryResultHandle) error {
+	if tq == nil || tq.max == 0 {
+		return ErrNoAvailableSharders
+	}
+
+	urls := make([]string, 0, tq.max)
+	for _, host := range tq.sharders {
+		urls = append(urls, tq.buildUrl(host, query))
+	}
+
+	r := resty.New()
+	r.DoGet(ctx, urls...).
+		Then(func(req *http.Request, resp *http.Response, respBody []byte, cf context.CancelFunc, err error) error {
+			res := QueryResult{
+				Content:    respBody,
+				Error:      err,
+				StatusCode: http.StatusBadRequest,
+			}
+			Logger.Debug(req.URL.String() + " " + resp.Status)
+
+			if resp != nil {
+				res.StatusCode = resp.StatusCode
+				Logger.Debug(string(respBody))
+			}
+
+			if handle != nil {
+				if handle(res) {
+
+					cf()
```

Comment:
Seems the calling of cf() actually does not cancel anything. All requests will be sent out anyway. Please check the r.Wait() below. Otherwise the function name should be FromAll().

Reply:
`ctx, cf := context.WithCancel`, so cf will not stop to send requests, but it will send cancel signal.